File: python/scripts/regenerate_examples.py
#!/usr/bin/env python3
"""Regenerate example files from Jinja2 templates."""

import re
import subprocess
from dataclasses import dataclass
from itertools import combinations
from pathlib import Path

import yaml
from jinja2 import Environment, FileSystemLoader


@dataclass
class ExampleTemplate:
    """Configuration for generating examples from a template."""

    path: Path
    basename: str
    options: list[str] | None  # e.g., ["async", "stream", "tools", "context"]


def parse_template_metadata(template_path: Path) -> ExampleTemplate:
    """Parse metadata from a Jinja2 template file."""
    content = template_path.read_text()

    # Look for metadata comment block
    pattern = r"\{\#\s*TEMPLATE_METADATA:(.*?)\#\}"
    match = re.search(pattern, content, re.DOTALL)

    if not match:
        raise ValueError(f"No TEMPLATE_METADATA found in {template_path}")

    metadata_yaml = match.group(1).strip()
    metadata = yaml.safe_load(metadata_yaml)

    # Extract basename from filename (remove .j2 extension)
    basename = template_path.stem

    return ExampleTemplate(
        path=template_path.parent,
        basename=basename,
        options=metadata.get("options", []),
    )


def discover_templates(root_path: Path) -> list[ExampleTemplate]:
    """Discover all *.j2 template files and parse their metadata."""
    templates = []
    for template_file in root_path.rglob("*.j2"):
        # Skip macro/include files (starting with _)
        if template_file.name.startswith("_"):
            continue

        template_config = parse_template_metadata(template_file)
        templates.append(template_config)

    return templates


def get_all_combinations(options: list[str] | None) -> list[list[str]]:
    """Get all valid combinations of options."""
    if options is None:
        options = []

    all_combos = []

    # Base case (no options)
    all_combos.append([])

    # All possible combinations of options
    for r in range(1, len(options) + 1):
        for combo in combinations(options, r):
            all_combos.append(list(combo))

    return all_combos


def get_filename(basename: str, combo: list[str]) -> str:
    """Get filename for a given combination of options."""
    if not combo:
        return f"{basename}.py"

    parts = [basename] + sorted(combo)
    return f"{'_'.join(parts)}.py"


def get_template_vars(options: list[str] | None, combo: list[str]) -> dict:
    """Get template variables for a given combination of options."""
    if options is None:
        options = []
    return {f"use_{option}": option in combo for option in options}


def generate_examples(template_config: ExampleTemplate) -> None:
    """Generate all example files for a template configuration."""
    # Remove existing Python files that match the basename pattern
    pattern = f"{template_config.basename}*.py"
    for existing_file in template_config.path.glob(pattern):
        existing_file.unlink()

    # Generate all combinations of options
    combinations = get_all_combinations(template_config.options)
    examples_root = Path(__file__).parent.parent / "examples"
    rel_path = template_config.path.relative_to(examples_root)
    template_path = rel_path / f"{template_config.basename}.j2"

    for combo in combinations:
        filename = get_filename(template_config.basename, combo)
        template_vars = get_template_vars(template_config.options, combo)

        # Create fresh environment for each template with global variables
        env = Environment(loader=FileSystemLoader(examples_root))
        env.globals.update(template_vars)
        template = env.get_template(str(template_path))

        content = template.render(**template_vars)
        filepath = template_config.path / filename

        with open(filepath, "w") as f:
            f.write(content)

    # Format all generated files with ruff
    generated_files = list(template_config.path.glob(f"{template_config.basename}*.py"))
    if generated_files:
        python_dir = Path(__file__).parent.parent
        subprocess.run(
            ["uv", "run", "ruff", "check", "--fix"] + [str(f) for f in generated_files],
            check=True,
            cwd=python_dir,
        )
        subprocess.run(
            ["uv", "run", "ruff", "format"] + [str(f) for f in generated_files],
            check=True,
            cwd=python_dir,
        )


def main() -> None:
    """Main entry point for regenerating examples."""
    examples_root = Path(__file__).parent.parent / "examples"
    templates = discover_templates(examples_root)

    for template_config in templates:
        generate_examples(template_config)


if __name__ == "__main__":
    main()

File: python/scripts/regenerate_examples.ts
#!/usr/bin/env bun
/**
 * Regenerate example files from TypeScript generator.
 */

import { writeFileSync, readdirSync, unlinkSync } from "fs";
import { join, dirname, relative } from "path";
import { spawn } from "child_process";
import {
  ExampleGenerator,
  filenameForOptions,
  allOptions,
} from "./example_generator.ts";

async function runCommand(
  command: string,
  args: string[],
  cwd: string
): Promise<void> {
  return new Promise((resolve, reject) => {
    const child = spawn(command, args, { cwd, stdio: "inherit" });
    child.on("close", (code) => {
      if (code === 0) {
        resolve();
      } else {
        reject(new Error(`Command failed with exit code ${code}`));
      }
    });
  });
}

async function main(): Promise<void> {
  const scriptPath = import.meta.path;
  const outputDir = join(dirname(scriptPath), "..", "examples", "sazed");

  // Remove existing Python files
  const files = readdirSync(outputDir);
  for (const file of files) {
    if (file.match(/^sazed.*\.py$/)) {
      unlinkSync(join(outputDir, file));
    }
  }

  // Generate all combinations
  const generatedFiles: string[] = [];
  for (const options of allOptions()) {
    const filename = filenameForOptions(options);

    const generator = new ExampleGenerator(options);
    const content = generator.generate();
    const filepath = join(outputDir, filename);

    writeFileSync(filepath, content);
    generatedFiles.push(filepath);
  }

  // Check formatting with ruff (without fixing)
  if (generatedFiles.length > 0) {
    const pythonDir = join(dirname(scriptPath), "..");
    try {
      await runCommand(
        "uv",
        ["run", "ruff", "check", "--fix", ...generatedFiles],
        pythonDir
      );
      await runCommand(
        "uv",
        ["run", "ruff", "format", ...generatedFiles],
        pythonDir
      );
    } catch (error) {
      console.error(
        "Ruff check failed! Generated code needs formatting fixes."
      );
      throw error;
    }
  }

  // Generate MDX file with all generated examples
  const docsRoot = join(dirname(scriptPath), "..", "..", "docs");
  const outputPath = join(docsRoot, "content", "examples.mdx");
  generateExamplesMdx(
    generatedFiles,
    join(dirname(scriptPath), "..", "examples"),
    outputPath
  );

  console.log(`Generated ${generatedFiles.length} example files`);
}

function generateExamplesMdx(
  generatedFiles: string[],
  examplesRoot: string,
  outputPath: string
): void {
  const content: string[] = [];
  content.push("---");
  content.push("title: Examples");
  content.push(
    "description: Unified doc with diverse examples, intended as a primer for LLMs on common Mirascope patterns."
  );
  content.push("---");
  content.push("");
  content.push("# Mirascope v2: Progressive Examples");
  content.push("");
  content.push(
    "The following examples show how to use Mirascope v2 with a variety of usage patterns."
  );
  content.push(
    "These are intended to give LLMs a primer on Mirascope's interface."
  );
  content.push("");

  for (const pyFile of generatedFiles) {
    const relPath = relative(examplesRoot, pyFile);
    content.push(`## \`${relPath}\``);
    content.push("");
    content.push(`<CodeExample file="examples/${relPath}" />`);
    content.push("");
  }

  writeFileSync(outputPath, content.join("\n"));
}

if (import.meta.main) {
  main().catch(console.error);
}
